transformations/tense/transformation.py

```diff
@@ -0,0 +1,167 @@
+import spacy
+from spacy.symbols import NOUN
+import random
+
+SUBJ_DEPS = {'agent', 'csubj', 'csubjpass', 'expl', 'nsubj', 'nsubjpass'}
+
+def _get_conjuncts(tok):
+    """
+    Return conjunct dependents of the leftmost conjunct in a coordinated phrase,
+    e.g. "Burton, [Dan], and [Josh] ...".
+    """
+    return [right for right in tok.rights
+            if right.dep_ == 'conj']
+
+
+def is_plural_noun(token):
+    """
+    Returns True if token is a plural noun, False otherwise.
+    Args:
+        token (``spacy.Token``): parent document must have POS information
+    Returns:
+        bool
+    """
+    if token.doc.is_tagged is False:
+        raise ValueError('token is not POS-tagged')
+    return True if token.pos == NOUN and token.lemma != token.lower else False
+
+
+def get_subjects_of_verb(verb):
+    if verb.dep_ == "aux" and list(verb.ancestors):
+        return get_subjects_of_verb(list(verb.ancestors)[0])
+    """Return all subjects of a verb according to the dependency parse."""
+    subjs = [tok for tok in verb.lefts if tok.dep_ in SUBJ_DEPS]
+    # get additional conjunct subjects
+    subjs.extend(tok for subj in subjs for tok in _get_conjuncts(subj))
+    if not len(subjs):
+        ancestors = list(verb.ancestors)
+        if len(ancestors) > 0:
+            return get_subjects_of_verb(ancestors[0])
+    return subjs
+
+
+def is_plural_verb(token):
+    if token.doc.is_tagged is False:
+        raise ValueError('token is not POS-tagged')
+    subjects = get_subjects_of_verb(token)
+    if not len(subjects):
+        return False
+    plural_score = sum([is_plural_noun(x) for x in subjects])/len(subjects)
+
+    return plural_score > .5
+
+def preserve_caps(word, newWord):
+    """Returns newWord, capitalizing it if word is capitalized."""
+    if word[0] >= 'A' and word[0] <= 'Z':
+        newWord = newWord.capitalize()
+    return newWord
+
+'''
+change tense function borrowed from https://github.com/bendichter/tenseflow/blob/master/tenseflow/change_tense.py
+'''
+
+class TenseTransformation(SentenceOperation):
+    tasks = [
+        TaskType.TEXT_CLASSIFICATION,
+        TaskType.TEXT_TO_TEXT_GENERATION,
+        TaskType.TEXT_TAGGING,
+    ]
+    languages = ["en"]
+
+    def __init__(self, to_tense):
+        super().__init__()
+        assert to_tense in ['past', 'present', 'future', 'random']
+        self.to_tense = to_tense
+        self.nlp = spacy.load('en_core_web_sm')
+
+    def change_tense(self, text, to_tense):
+        """Change the tense of text.
+        Args:
+            text (str): text to change.
+            to_tense (str): 'present','past', or 'future'
+            npl (SpaCy model, optional):
+        Returns:
+            str: changed text.
+        """
+        tense_lookup = {'future': 'inf', 'present': PRESENT, 'past': PAST}
+        tense = tense_lookup[to_tense]
+
+        doc = self.nlp(text)
+
+        out = list()
+        out.append(doc[0].text)
```

Comment:
Why not change the tense of all sentences?

Reply:
There seems to be a misunderstanding , we are indeed changing the tense of all sentences, here the variable `out` is just initialised with the first word(token) in the given sentence, a few lines below `for word in doc` we iterate over all words in the sentence